### PR TITLE
Chatbot request access

### PR DESCRIPTION
**Pull request by pgobin**
### Auth:
  - No longer possible to retrieve JWT from JS
  - Websocket handshake secured
  - Webhook endpoint secured using JWT sent in Rasa slot
  - **Note:** CSRF not addressed in this PR
  
### CORS:
  - Added CORS sugar used for per-resource policy enforcement
  - API CORS policy disallows `*`
  - **Note:** `CLIENT_HOST` passed to CORS is for connection from outside Docker

### Webhook:
  - Prevent requests from blocking event loop
  - Add ability for webhook to request role membership

### Diagram:

Signed-off-by: PGobz <p.gobin@gmail.com>

### CODE REVIEW

File: rbac/server/api/utils.py [hunk 1]
```diff
@@ -39,6 +42,29 @@ def validate_fields(required_fields, body):
         raise ApiBadRequest("Bad Request: Improper JSON format")
 
 
+def create_authorization_response(token, data):
+    response = json({"data": data})
```
Comment (nadiabahrami) [inline rbac/server/api/utils.py]: Need a docstring explaining function

File: rbac/server/api/utils.py [hunk 2]
```diff
@@ -39,6 +42,29 @@ def validate_fields(required_fields, body):
...
+    response.cookies[SIGNATURE_KEY] = ".".join(token.split(".")[0:2])
+    response.cookies[PAYLOAD_KEY] = token.split(".")[2]
+
+    response.cookies[SIGNATURE_KEY]["httponly"] = True
+    return response
+
+
+def extract_request_token(request):
+    try:
```
Comment (nadiabahrami) [inline rbac/server/api/utils.py]: Docstring

File: rbac/server/api/utils.py [hunk 3]
```diff
@@ -39,6 +42,29 @@ def validate_fields(required_fields, body):
...
+        return request.json["tracker"]["slots"]["token"]
+    except (KeyError, TypeError):
+        pass
+    if request.cookies.get(SIGNATURE_KEY) is None:
```
Comment (nadiabahrami) [inline rbac/server/api/utils.py]: Are you trying to check if it is None or an empty string here?

File: rbac/server/api/webhooks.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+from rbac.server.api import utils
+
+LOGGER = logging.getLogger(__name__)
+WEBHOOKS_BP = Blueprint("webhooks")
+
+
+@WEBHOOKS_BP.post("webhooks/chatbot")
+@authorized()
+async def chatbot_webhook(request):
+    required_fields = ["tracker", "sender_id"]
```
Comment (nadiabahrami) [inline rbac/server/api/webhooks.py]: docstrings on all new functions

File: rbac/server/api/utils.py (hunk 1 above)
Comment (pgobin) [reply rbac/server/api/utils.py]: # 👾

File: rbac/server/api/utils.py (hunk 3 above)
Comment (pgobin) [reply rbac/server/api/utils.py]: None.

File: rbac/server/api/webhooks.py (hunk 4 above)
Comment (pgobin) [reply rbac/server/api/webhooks.py]: # 👾

File: rbac/server/api/utils.py (hunk 2 above)
Comment (pgobin) [reply rbac/server/api/utils.py]: # 👾

**Pull request merged**